docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,153 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Node Allocatable Resources
+
+**Issue:** https://github.com/kubernetes/kubernetes/issues/13984
+
+## Overview
+
+Currently Node.Status has Capacity, but no concept of Machine Allocatable. We need additional
+parameters to serve several purposes:
+
+1. [Kubernetes metrics](compute-resource-metrics-api.md) provides "/docker-daemon", "/kubelet",
+   "/kube-proxy", "/system" etc. raw containers for monitoring system component resource usage
+   patterns and detecting regressions. Eventually we want to cap system component usage to a certain
+   limit / request. However this is not currently feasible because 1) docker still uses tons of
+   computing resources and 2) there is no node config yet, so we cannot control Kubenetes nodes or know
+   about arbitrary daemons which can make /system unmanageable. Even with a resource cap we cannot
+   do a full resource management on the node, but with the proposed parameters we can mitigate
+   really bad resource overcommits.
+2. For external schedulers (such as mesos, hadoop, etc.) integration, they might want to partition
+   compute resources on a given node, limiting how much Kubelet can use. We should provide a
+   mechanism by which they can query kubelet, and reserve some resources for their own purpose.
+
+## Design
+
+### Definitions
+
+![image](node-allocatable.png)
+
+1. **Machine Capacity** - Already provided as
+   [`NodeStatus.Capacity`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus),
+   this is total capacity read from the machine instance, and assumed to be constant.
+2. **Reserved** (proposed) - Compute resources reserved for non-Kubernetes and non-system
+   workloads. For example, processes scheduled to run alongside Kubernetes on Mesos.
+3. **Kubelet Allocatable** - Compute resources available for scheduling (including scheduled &
+   unscheduled resources). This value is currently implicit, and the focus of this proposal. See
+   [below](#api-changes) for more details.
+4. **System Reserved** (proposed) - Compute resources reserved for system components such as the docker daemon, kubelet, kube proxy, etc.
+
+### API changes
+
+#### Allocatable
+
+Add `Allocatable` (4) to
+[`NodeStatus`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus):
+
+```
+type NodeStatus struct {
+  ...
+  // Allocatable represents schedulable resources of a node.
+  Allocatable ResourceList `json:"allocatable,omitempty"`
+  ...
+}
+```
+
+Allocatable will be computed by the Kubelet and reported periodically to the API server. It is
+defined to be:
+
+```
+   [Allocatable] = [Machine Capacity] - ( [Reserved] + [System Reserved] + [Kernel usage (smoothed)] )
```

Comment:
@bgrant0607 We eventually want to enforce limits based on `SystemReserved`, but since we can't control the kernel usage we don't want to include the kernel in SystemReserved. I suppose we could include it in `Reserved`, but that seems like overloading the concept for something it wasn't intended for, and also runs the risk of the kernel running over the limit.